File: DarkKitchen.BusinessLogic/ProductService.cs
using DarkKitchen.Domain.Products;
using DarkKitchen.IBusinessLogic;
using DarkKitchen.IDataAccess;
using DarkKitchen.Models.Converters;
using DarkKitchen.Models.DTOs;

namespace DarkKitchen.BusinessLogic;

public class ProductService(IProductRepository productRepository) : IProductService
{
    private readonly IProductRepository _productRepository = productRepository;

    public IEnumerable<ProductResponse> GetProducts(string? name, string? line, string? category)
    {
        var products = _productRepository.GetAll();

        if(!string.IsNullOrWhiteSpace(name))
        {
            products = products.Where(p => p.Name.Contains(name, StringComparison.OrdinalIgnoreCase));
        }

        if(!string.IsNullOrWhiteSpace(line))
        {
            products = products.Where(p => p.Line.Name.Equals(line, StringComparison.OrdinalIgnoreCase));
        }

        if(!string.IsNullOrWhiteSpace(category))
        {
            products = products.Where(p => p.Category.Name.Equals(category, StringComparison.OrdinalIgnoreCase));
        }

        return products.Select(Converter.ToProductResponse);
    }

    public ProductResponse CreateProduct(ProductCreateRequest request)
    {
        var product = Converter.ToProduct(request);
        _productRepository.Add(product);
        return Converter.ToProductResponse(product);
    }

    public ProductResponse UpdateProduct(Guid id, ProductUpdateRequest request)
    {
        Product product = _productRepository.GetById(id)
                          ?? throw new KeyNotFoundException($"Producto {id} no encontrado.");

        var line = new ProductLine(request.Line);
        var category = new ProductCategory(request.Category);
        var images = request.Images
            .Select(i => new ProductImage(i.Url, i.SizeInBytes))
            .ToList();

        product.UpdateDetails(request.Name, request.Description, line, category, request.Price, images);

        if(request.IsActive.HasValue)
        {
            if(request.IsActive.Value)
            {
                product.Activate();
            }
            else
            {
                product.Deactivate();
            }
        }

        _productRepository.Update(id, product);
        return Converter.ToProductResponse(product);
    }
}

File: DarkKitchenUI/src/app/features/products/services/product.service.ts
import { Injectable, signal } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import {
  ProductResponse,
  ProductCreateRequest,
  ProductUpdateRequest,
} from '../models/product.models';

@Injectable({
  providedIn: 'root',
})
export class ProductService {
  private apiUrl = `${environment.apiUrl}/products`;

  products = signal<ProductResponse[]>([]);
  isLoading = signal(false);

  constructor(private http: HttpClient) {}

  getAll(name?: string, line?: string, category?: string): Observable<ProductResponse[]> {
    let params = new HttpParams();
    if (name) params = params.set('name', name);
    if (line) params = params.set('line', line);
    if (category) params = params.set('category', category);

    return this.http.get<ProductResponse[]>(this.apiUrl, { params });
  }

  create(request: ProductCreateRequest): Observable<ProductResponse> {
    return this.http.post<ProductResponse>(this.apiUrl, request);
  }

  update(id: string, request: ProductUpdateRequest): Observable<ProductResponse> {
    return this.http.put<ProductResponse>(`${this.apiUrl}/${id}`, request);
  }

  toggleActive(product: ProductResponse): Observable<ProductResponse> {
    const request: ProductUpdateRequest = {
      name: product.name,
      description: product.description,
      line: product.line,
      category: product.category,
      price: product.price,
      images: product.images.map((img) => ({
        url: img.url,
        sizeInBytes: img.sizeInBytes,
      })),
      isActive: !product.isActive,
    };
    return this.update(product.id, request);
  }
}
